perun-oidc-server/src/main/java/cz/muni/ics/oauth2/web/DeviceEndpoint.java

```diff
@@ -164,138 +210,230 @@ public String requestDeviceCode(@RequestParam("client_id") String clientId, @Req
+		ClientDetailsEntity client;
+
+		try {
+			client = clientService.loadClientByClientId(clientId);
+			Collection<String> authorizedGrantTypes = client.getAuthorizedGrantTypes();
+
+			if (authorizedGrantTypes != null
+					&& !authorizedGrantTypes.isEmpty()
+					&& !authorizedGrantTypes.contains(DeviceTokenGranter.GRANT_TYPE)) {
+				throw new InvalidClientException("Unauthorized grant type: " + DeviceTokenGranter.GRANT_TYPE);
+			}
+
+		} catch (IllegalArgumentException e) {
+			log.error("IllegalArgumentException was thrown when attempting to load client", e);
+			model.put(HttpCodeView.CODE, HttpStatus.BAD_REQUEST);
+			return HttpCodeView.VIEWNAME;
+
+		} catch (OAuth2Exception e) {
+			log.error("could not find client " + clientId);
+			model.put(HttpCodeView.CODE, HttpStatus.NOT_FOUND);
+			return HttpCodeView.VIEWNAME;
+		}
+
+		// make sure the client is allowed to ask for those scopes
+		Set<String> requestedScopes = OAuth2Utils.parseParameterList(scope);
+		Set<String> allowedScopes = client.getScope();
+
+		if (!scopeService.scopesMatch(allowedScopes, requestedScopes)) {
+			log.error("Client asked for " + requestedScopes + " but is allowed " + allowedScopes);
+			model.put(HttpCodeView.CODE, HttpStatus.BAD_REQUEST);
+			model.put(JsonErrorView.ERROR, INVALID_SCOPE);
+
+			return JsonErrorView.VIEWNAME;
+		}
+
+		clientWithScopes.setClient(client);
+		clientWithScopes.setRequestedScopes(requestedScopes);
+
+		return null;
+	}
+
+	private String checkDeviceCodeIsValid(DeviceCode dc, HttpServletRequest req, ModelMap model) {
+		// we couldn't find the device code
+		if (dc == null) {
+			model.addAttribute(ERROR, NO_USER_CODE);
+			return tryToSetThemedRequestUserCode(req, model);
+		}
+
+		// make sure the code hasn't expired yet
+		if (dc.getExpiration() != null && dc.getExpiration().before(new Date())) {
+			model.addAttribute(ERROR, EXPIRED_USER_CODE);
+			return tryToSetThemedRequestUserCode(req, model);
+		}
+
+		// make sure the device code hasn't already been approved
+		if (dc.isApproved()) {
+			model.addAttribute(ERROR, USER_CODE_ALREADY_APPROVED);
+			return tryToSetThemedRequestUserCode(req, model);
+		}
+
+		return null;
+	}
+
+	private String tryToSetThemedRequestUserCode(HttpServletRequest req, ModelMap model) {
+		if (!perunOidcConfig.getTheme().equalsIgnoreCase(DEFAULT)) {
+			ControllerUtils.setPageOptions(model, req, htmlClasses, perunOidcConfig);
+			model.put(PAGE, REQUEST_USER_CODE);
+			String shibAuthnContextClass = "";
+
+			if (StringUtils.hasText(req.getParameter(ACR_VALUES))) {
+				shibAuthnContextClass = (String) req.getAttribute(PerunFilterConstants.SHIB_AUTHN_CONTEXT_CLASS);
+
+				if (!StringUtils.hasText(shibAuthnContextClass)) {
+					shibAuthnContextClass = (String) req.getAttribute(PerunFilterConstants.SHIB_AUTHN_CONTEXT_METHOD);
+				}
+
+				if (!StringUtils.hasText(shibAuthnContextClass)) {
+					shibAuthnContextClass = "";
+				}
+			}
+
+			model.put(ACR, shibAuthnContextClass);
```

Comment:
This is actually not necessary anymore. As we have switched to SAML SP integration (instead of having Shibboleth SP in front of the OP) the information is now available elsewhere and will be set in the authentication context and this whole block is thus not needed.